src/twisted/test/ssl_helpers.py

```diff
@@ -18,21 +18,38 @@
 
 
 class ClientTLSContext(ssl.ClientContextFactory):
+    """
+    SSL Context Factory for client-side connections.
+    """
+
     isClient = 1
+    _context = None
 
     def getContext(self):
-        return SSL.Context(SSL.SSLv23_METHOD)
+        if self._context is None:
+            self._context = SSL.Context(SSL.SSLv23_METHOD)
+        return self._context
 
 
 class ServerTLSContext:
+    """
+    SSL Context Factory for server-side connections.
+    """
+
     isClient = 0
+    _context = None
 
-    def __init__(self, filename=certPath, method=SSL.SSLv23_METHOD):
+    def __init__(self, filename=certPath, method=None):
         self.filename = filename
+        if method is None:
+            method = SSL.SSLv23_METHOD
+
         self._method = method
 
     def getContext(self):
-        ctx = SSL.Context(self._method)
-        ctx.use_certificate_file(self.filename)
-        ctx.use_privatekey_file(self.filename)
-        return ctx
+        if self._context is None:
+            self._context = SSL.Context(self._method)
+            self._context.use_certificate_file(self.filename)
+            self._context.use_privatekey_file(self.filename)
+
+        return self._context
```

Comment:
Are any of these caching changes actually necessary?  It doesn't seem like they should be.

Ahh.  I guess this was an attempt to fix the ALPN issue.  It doesn't seem like it should be necessary and I don't think we can *force* applications to cache their context anyway.  So we should probably back this out and find another solution to the failing ALPN test.

Reply:
No.
They should be reverted once the fix this... or even sooner.

Thanks for the headup...as I would have most probably forgot about this.

I left them there to help with debugging. As far as I remember, in some places I only has a reference to the factory and not to the context...so I cached them to make it easy to retrieve them, without adding other helpers in the implementation code.

I think that is important not to cache them during tests as it helps detect implementation issues.
So maybe add a comment to this implementation that the context are not cached on purpose.

One of my guess was that the connection is not set up with the same context...so I have updated this code here.

But yes. This should be reverted.

I am curious if there is any production server which uses a different context for each connection.

I guess that you can have an ACME (Let's Encrypt) context factory that will automatically update the cert in the background. The new connections can get a refreshed certificate without the need to restart the port. 


I can imagine that you can have a server that will server one certificate if the source is an internal IP and another certificate if is an external IP ... but the current API does not support this use case, as getContext() is called without knowing the peer that will handle TLS